feat: add route for force crawling

bolinocroustibat opened the pull request:
Closes https://github.com/datagouv/data.gouv.fr/issues/1395 (partly).

This ticket has been split into 5 PRs for easier reviewing:
1) Refactoring DB methods for new routes: [PR #112](https://github.com/datagouv/hydra/pull/112) [✅](https://emojipedia.org/check-mark-button)
2) Refactoring of old routes [PR #117](https://github.com/datagouv/hydra/pull/117) [✅](https://emojipedia.org/check-mark-button)
3) Adding new route to gather informations from a resource in the catalog (ex : url, priority, etc.) [PR #120] [✅](https://emojipedia.org/check-mark-button)
4) **This PR**: Adding new route to force a crawl [PR #118](https://github.com/datagouv/hydra/pull/118)📍
5) Adding resource statuses and a resource status route: [PR #119](https://github.com/datagouv/hydra/pull/119)

**Details of this PR:**
- Create a new POST route `/api/checks/create/` to force a new crawl on a specific `resource_id` or `url`. _Note: no authentification in this PR yet, I think we should do it, but in another PR as this one is already big_
- Create a new related test `tests/test_api.py::test_api_resource_created`

magopian commented on tests/test_api.py:
```diff
@@ -94,8 +94,31 @@ async def test_api_get_all_checks(setup_catalog, client, query, fake_check):
 
 
-async def test_api_get_resource(db, client, insert_fake_resource):
-    await insert_fake_resource(db)
```
The diff in github is a bit weird, but it makes it look like this line is gone? Is that on purpose?

magopian reviewed (approved):
Looks good to me :+1:

Do you want to add that to the CHANGELOG? (maybe we need a CI target that automatically checks the CHANGELOG has been updated... or even a tool that automatically populates the CHANGELOG from the commits ? To be discussed ;)

Pierlou commented on udata_hydra/routes/checks.py:
```diff
@@ -22,3 +28,26 @@ async def get_all_checks(request: web.Request) -> web.Response:
+    """Create a new check"""
+    try:
+        payload: dict = await request.json()
+        url = payload["url"]
```
I find it weird to ask for both the resource's URL and id, and I guess it could be error-prone (what happens if you mismatch URL and id?). Wouldn't there be a way to manage with only the id? We could get the URL from the db or data.gouv directly 🤔

Pierlou commented on tests/test_api.py:
```diff
@@ -94,8 +94,31 @@ async def test_api_get_all_checks(setup_catalog, client, query, fake_check):
+
+
+async def test_api_create_check(setup_catalog, client, fake_check):
+    await fake_check(status=500, error="no-can-do")
```
I'm guessing this will go eventually and was left as a reminder?

Pierlou commented on udata_hydra/routes/checks.py:
```diff
@@ -22,3 +28,26 @@ async def get_all_checks(request: web.Request) -> web.Response:
+        async with aiohttp.ClientSession(
+            timeout=None, headers={"user-agent": config.USER_AGENT}
+        ) as session:
+            result = await check_url(url=url, resource_id=resource_id, session=session)
```
Wondering if in this specific case we'd like the check to have high priority. Currently `check_url` [always sets priority as low](https://github.com/datagouv/hydra/blob/46b73144792392557255a72f3ee63ed88bf2045f/udata_hydra/crawl.py#L288C27-L288C43). Thoughts?

bolinocroustibat commented:
> Looks good to me 👍
> 
> Do you want to add that to the CHANGELOG? (maybe we need a CI target that automatically checks the CHANGELOG has been updated... or even a tool that automatically populates the CHANGELOG from the commits ? To be discussed ;)

Good point, I forgot. Changed now and also fixed a fe typos in the changelog.

I would be very interested by the second option, to populate the changelog automatically from commit names and descriptions. That would be a very big gain in productivity, and it would force us to do even better commits names and descriptions :)

Pierlou reviewed (approved):
Looking forwards to having this in production!

bolinocroustibat commented:
I will wait for @maudetes to review the whole logic (especially the priority logic) to make sure it's ok, before merging.
So keeping this PR open for a few more days before merging :)

maudetes commented on udata_hydra/cli.py:
```diff
@@ -144,7 +144,9 @@ async def check_resource(resource_id: str, method: str = "get"):
     async with aiohttp.ClientSession(timeout=None) as session:
-        await crawl_check_url(url=res[0], resource_id=None, session=session, method=method)
+        await crawl_check_url(
+            url=res[0], resource_id=None, session=session, method=method, worker_priority="low"
```
Actually, since it's quite an equivalent from the API call, we may want to have `worker_priority="high"` as well in the cli.

maudetes commented on udata_hydra/crawl.py:
```diff
@@ -235,7 +237,12 @@ def has_nice_head(resp) -> bool:
+    session,
+    sleep: float = 0,
+    method: str = "head",
+    worker_priority: str = "default",
```
Thanks for the renaming, it is clearer indeed!

maudetes reviewed (approved):
Thank yo!

I am thinking of future iterations (running the different tasks inline instead of going through a worker, an additional priority level, etc.) but we can start with this version for a starter!

It would be wiser to test this new endpoint for real usage examples.
